DirectProgramming/DPC++/N-BodyMethods/Nbody/src/GSimulation.cpp

```diff
@@ -0,0 +1,234 @@
+// Copyright © 2020 Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+
+// =============================================================
+// Copyright (c) 2019 Fabio Baruffa
+// Source: https://github.com/fbaru-dev/particle-sim
+// MIT License
+// =============================================================
+
+#include "GSimulation.hpp"
+#include "dpc_common.hpp"
+using namespace sycl;
+using namespace dpc_common;
+
+/* Default Constructor for the GSimulation class which sets up the default values for 
+ * number of particles, number of integration steps, time steo and sample frequency */
+GSimulation::GSimulation() {
+  std::cout << "===============================" << "\n";
+  std::cout << " Initialize Gravity Simulation" << "\n";
+  set_npart(16000);
+  set_nsteps(10);
+  set_tstep(0.1);
+  set_sfreq(1);
+}
+
+/* Set the number of particles */
+void GSimulation::SetNumberOfParticles(int N) { set_npart(N); }
+
+/* Set the number of integration steps */
+void GSimulation::SetNumberOfSteps(int N) { set_nsteps(N); }
+
+/* Initialize the position of all the particles using random number generator between 0 and 1.0 */
+void GSimulation::InitPos() {
+  std::random_device rd;  // random number generator
+  std::mt19937 gen(42);
+  std::uniform_real_distribution<RealType> unif_d(0, 1.0);
+
+  for (int i = 0; i < get_npart(); ++i) {
+    particles_[i].pos[0] = unif_d(gen);
+    particles_[i].pos[1] = unif_d(gen);
+    particles_[i].pos[2] = unif_d(gen);
+  }
+}
+
+/* Initialize the velocity of all the particles using random number generator between -1.0 and 1.0 */
+void GSimulation::InitVel() {
+  std::random_device rd;  // random number generator
+  std::mt19937 gen(42);
+  std::uniform_real_distribution<RealType> unif_d(-1.0, 1.0);
+
+  for (int i = 0; i < get_npart(); ++i) {
+    particles_[i].vel[0] = unif_d(gen) * 1.0e-3f;
+    particles_[i].vel[1] = unif_d(gen) * 1.0e-3f;
+    particles_[i].vel[2] = unif_d(gen) * 1.0e-3f;
+  }
+}
+
+/* Initialize the acceleration of all the particles to 0 */
+void GSimulation::InitAcc() {
+  for (int i = 0; i < get_npart(); ++i) {
+    particles_[i].acc[0] = 0.f;
+    particles_[i].acc[1] = 0.f;
+    particles_[i].acc[2] = 0.f;
+  }
+}
+
+/* Initialize the mass of all the particles using a random number generator between 0 and 1 */
+void GSimulation::InitMass() {
+  RealType n = static_cast<RealType>(get_npart());
+  std::random_device rd;  // random number generator
+  std::mt19937 gen(42);
+  std::uniform_real_distribution<RealType> unif_d(0.0, 1.0);
+
+  for (int i = 0; i < get_npart(); ++i) {
+    particles_[i].mass = n * unif_d(gen);
+  }
+}
+
+/* This function does the simulation logic for Nbody */
+void GSimulation::Start() {
+  RealType dt = get_tstep();
+  int n = get_npart();
+  RealType* energy = new RealType[n];
+  for (int i = 0; i < n; i++) energy[i] = 0.f;
+  // allocate particles
```

Comment:
Idem.
You can then remove `GSimulation::~GSimulation()`
The destructor was wrong anyway.
So please do not use pointers and memory allocation in a tutorial without any serious reason.
There were 3 at least 3 bugs in this code, so imagine the ethical impact on C++ beginners or Python programmers reading this... :-(

Reply:
Not using pointers anymore. Switched to std::vector<> in the latest code